### PR TITLE
Rebuild with `HList` instead of `Tuple`

### PR DESCRIPTION
**Pull request by mrdziuban**
Defines an `HList` type and uses it in place of `Tuple`. This was prompted by the performance characteristics of `Tuple`s, specifically the performance of head/tail decomposition, which is what the vast majority of this library is built on.

Benchmarks of shapeless `HList` in Scala 2 vs. `Tuple` in Scala 3: https://github.com/mrdziuban/dotty-tuple-shapeless-benchmark

Discussion with Scala maintainers about why `Tuple` performance is worse: https://discord.com/channels/632150470000902164/632628489719382036/1143977364649431151

TL;DR
  - Scala 3 tuples are stored as instances of `TupleN` classes. When you call `.tail`, the elements have to be copied into a new instance.
  - Tuples are still primarily the good old tuples from Scala 2, with fast indexing. That of course is at odds with fast linear decomposition into a head and tail.

### CODE REVIEW

**Review by jeffdyke — COMMENTED** [review]
This is really `dope` @mrdziuban.  Been reading what you've clearly run into.  Nice work.
